### ISSUE TITLE
One click configuration

### ISSUE THREAD

**Issue by terma** [jenkinsci/ec2-fleet-plugin#144]
### AWS API too generic
Plugin user needs to figure a lot of out of scope before actual configuration. The plugin can significantly narrow this part by providing more specific options for configuration and even do some default preconfiguration which should reduce configuration time and simplify process

### AWS Console or AWS CLI require additional level of knoweledge
This knoweledge is out of scope for Jenkins users who just want to run their builds. Plugin could try to avoid dealing with actual AWS CLI/Console to streamline configuration and reduce knoweledge bar

### A lot of routine for Big Jenkins
Plugin requires configuration while it could be really small for one two fleets, it became big part of work if you have more than a few fleets as you need repeat same actions for each fleet. Previos two points should reduce this price, however plugin can reduce it even more.

## Improvements

- [ ] [One-click configuration](#one-click-configuration) #171 
- [ ] [Template Configuration](#template-configuration)
- [x] [Automatic Label based Configuration](#automatic-label-based-configuration) released in version ```1.17.0``` #160 

### One-click configuration

Auto create required resource during plugin configuration. The user just installs the plugin, type AWS credentials, select region, set cluster size and save
  - The plugin will create EC2 Spot Fleet with required IAM role and default instance type like ```C*``` 
  - All other configuration will be hidden behind ```Advanced``` button

### Template Configuration

Allow to create templates of fleet configuration, so they could be reused with some small overrides. This reduces amount of configuration. From other point of view it allows to limit user configuration by predefined parameters.

```Investigation in progress```

### Automatic Label based Configuration

Allow plugin user to configure plugin with credentials to create EC2 Spot Fleet and specify template settings. While actual configuration happens by labeling job with configuration string. Plugin will make sure that configuration from jobs label reflected as EC2 Spot Fleet.

For example, ```admin``` created template configuration to use ```c5``` and region ```us-east-1```, after that any user of plugin just need to label their jobs like:

```
ec2-fleet-plugin:ami=ami-00002c141987d373c;min=0;max=200;ondemand
```

This label will tell plugin to automatically create fleet based on template settings with requested ```ami, min, max, ondemand``` That's it. ```admin``` doesn't need to be involved.

#### Spot Fleet requires configuration parameters:
- AMI, name```aws-elasticbeanstalk-amzn-2019.02.13.x86_64-WindowsServer2012R2-V2-hvm-201903122020``` or ID ```ami-00002c141987d373c```
- Subnet ```vpc-000000``` or ```default```
- AZs ```us-east-1a```
- key pair ```my-key-pair```
- Instance Type ```c3.large```
- Spot/OnDemand
- Min/Max

#### Configuration Dialog Info
- https://github.com/jenkinsci/credentials-plugin/blob/43be901b61cc5d6345a2f97d6901067326045bc9/src/main/resources/lib/credentials/select.jelly
- https://github.com/jenkinsci/credentials-plugin/blob/6f710a7e30156e94aff8a3c6c6799cef04ad35df/src/main/resources/com/cloudbees/plugins/credentials/CredentialsSelectHelper/WrappedCredentialsStore/dialog.jelly
- https://github.com/jenkinsci/credentials-plugin/blob/6f710a7e30156e94aff8a3c6c6799cef04ad35df/src/main/resources/com/cloudbees/plugins/credentials/CredentialsSelectHelper/WrappedCredentialsStore/credential.jelly

**Comment by iponweb-github-admin** [comment]
Hi,
One-click configuration and Template Configuration really should improve the convenience of the plugin usage. Automatic Label based Configuration sounds like a killer-feature and it would be super nice to have it.

**Comment by terma** [comment]
@iponweb-github-admin thx for feedback, will work in that way =)

**Comment by ianfixes** [comment]
There is definitely value in having the plugin check whether it has proper AWS permissions to modify the spot request (for scaling up and down).  

Automatically creating the spot fleet requests (based on a AMI ID) would be very useful, but it should also allow
* input for the block device ID -- our build agents have a 150GB storage device attached
* Constraints on cost
* Tags for the instances
* Security groups

Would we be assured that the spot requests would be automatically cancelled when removed from the configuration?  How would we be assured that unrelated spot requests would not be deleted accidentally?

**Comment by terma** [comment]
@ianfixes thx for points, some answers https://github.com/jenkinsci/ec2-fleet-plugin/pull/160#issuecomment-568789903

**Comment by terma** [comment]
Label based base released under version [1.17.0](https://github.com/jenkinsci/ec2-fleet-plugin/releases/tag/ec2-fleet-1.17.0)
